Example Guide: Benchmark on Saved Models (#92)
aai-institute/nnbench: docs/tutorials/on_saved.md

# Benchmark on saved models
There is a high likelihood that you, at some point, find yourself wanting to benchmark models that were trained previously.
In this guide we will walk through how we can accomplish this with nnbench.
We will start with an aside that talks through the setup of the example we will use in this guide.

> If you are only interested in the application of nnbench, you can skip this section.

Reviewer: I would put these into the NER section below, and leave out the exposition (L4) or maybe reduce it to "This is a very short primer on the experiment setup".

## Example: Named Entity Recognition

> There are several reasons why you want to retrieve saved models for benchmarking. 

Reviewer: ```suggestion
There are lots reasons why you could want to retrieve saved models for benchmarking. 
```

> Among others, these are reviewing the work of colleagues, comparing experimental performances to a benchmark, or dealing with models that require significant compute such that in place retraining is impractical.

Reviewer: ```suggestion
Among them are reviewing the work of colleagues, comparing experimental performances to an existing benchmark, or dealing with models that require significant compute such that in-place retraining is impractical.
```

For this example, we deal with a named entity recognition (NER) model that is based on the pre-trained encoder-decoder transformer [BERT](https://arxiv.org/abs/1810.04805).
The model is trained on the [CoNLLpp dataset](https://huggingface.co/datasets/conllpp) which consists of sentences from news stories where words were tagged with Person, Organization, Location, or Miscellaneous if they referred to entities. 

> Words were assigned an out of entity label if they did not represented an entity.

Reviewer: ```suggestion
Words were assigned an out-of-entity label if they do not represent an entity.
```

### Model Training

> You find the code to train the model in the nnbench [repository](https://github.com/aai-institute/nnbench) in the directory `examples/on_saved/src/training/training.py`.

Reviewer: Let's adjust the folder name here and subsequently, too, if we decide to do so.

> If you want to skip running the training script but still want to reproduce this example, you can take any BERT model fine tuned for NER with the CoNLL- dataset family.

Reviewer: ```suggestion
If you want to skip running the training script but still want to reproduce this example, you can take any BERT model fine tuned for NER with the CoNLL dataset family.
```

> You find many on the Huggingface model hub, for example [this one](https://huggingface.co/dslim/bert-base-NER). You need to download the `model.safetensors`, `config.json`, `tokenizer_config.json`, and `tokenizer.json`.

Reviewer: ```suggestion
You find many on the Huggingface model hub, for example [this one](https://huggingface.co/dslim/bert-base-NER). You need to download the `model.safetensors`, `config.json`, `tokenizer_config.json`, and `tokenizer.json` files.
```

If you want to train your own model, continue below. 

> 

Reviewer: ```suggestion
```

When you open a new notebook there make sure to select a GPU instance in the upper right corner.
The you can upload the `training.py`.
You can ignore An eventual warning that the data is not persisted.
Next, install the necessary dependencies: `!pip install datasets transformers[torch]`.
Google Colab comes with some dependencies already installed in the environment.
Hence, if you are working with a different GPU instance, make sure to install everything from the `pyproject.toml` in the `examples/on_saved` folder. 
Next you can execute the `training.py` with `!python training.py`.
This will train two BERT models ("bert-base-uncased" and "distilbert-base-uncased") which we can compare using nnbench. 
If you want, you can adapt the training script to train other models by editing the tuples in the `tokenizers_and_models` list at the bottom of the training script. 
The training of the models takes around 10 minutes.
Once it is done, download the respective files and save them to your disk.
They should be the same mentioned above. 
We will need the path to the files for benchmarking later.

> 

Reviewer: ```suggestion
```

### The Benchmarks
The benchmarking code is found in the `examples/on_saved/benchmark.py`.
We calculate precision, recall, accuracy, and f1 scores for the whole test set and specific labels.
Additionally, we obtain information about the model such as its memory footprint and inference time.
We are not walking through the whole file but instead point out certain design choices as an inspiration to you. 

> If you are interested in a more detailed walkthrough of the setup of benchmarks, you find it [here](../guides/benchmarks.md).

Reviewer: ```suggestion
If you are interested in a more detailed walkthrough on how to set up benchmarks, you can find it [here](../guides/benchmarks.md).
```
